Closes #2541: x-axis improvements.

emtwo opened the pull request.

emtwo commented on client/app/services/query-result.js:
```diff
@@ -102,6 +102,9 @@ function QueryResultService($resource, $timeout, $q) {
             } else if (isString(v) && v.match(/^\d{4}-\d{2}-\d{2}$/)) {
               row[k] = moment.utc(v);
               newType = 'date';
+            } else if (v.match(/^\d{4}\d{2}\d{2}$/)) {
```
Perhaps this? `^\d{4}[:\/\-.]*\d{2}[:\/\-.]*\d{2}T*$`

kravets-levko replied:
`moment` can handle this properly, but this regex also matches strings like `2018-05.16`, and I'm not sure if it's okay.

emtwo replied:
hm. might be a more complicated regex statement to make sure both delimiters match... perhaps we can leave this as-is for now and have that in a separate issue?

emtwo replied:
https://github.com/getredash/redash/issues/2543

arikfr commented on client/app/services/query-result.js:
```diff
@@ -102,6 +102,9 @@ function QueryResultService($resource, $timeout, $q) {
             } else if (isString(v) && v.match(/^\d{4}-\d{2}-\d{2}$/)) {
               row[k] = moment.utc(v);
               newType = 'date';
+            } else if (v.match(/^\d{4}\d{2}\d{2}$/)) {
```
Unless I'm missing something, this regex has potential for a lot of false detections. Basically any numeric column in the range of 10000000-99999999 will be detected as dates, no?

emtwo replied:
Yes, you're right. I took the regex used above and removed the dashes but that makes the range of error worse than above. I've updated with a new regex pattern that is a little more restrictive.

arikfr commented on client/app/visualizations/chart/chart-editor.html:
```diff
@@ -159,9 +159,9 @@
-        <ui-select-match placeholder="Choose Scale...">{{$select.selected | capitalize}}</ui-select-match>
-        <ui-select-choices repeat="scaleType in xAxisScales">
-          <div ng-bind-html="scaleType | capitalize | highlight: $select.search"></div>
+        <ui-select-match placeholder="Choose Scale...">{{$select.selected.label | capitalize}}</ui-select-match>
```
The `label` field is already supposed to be human friendly string, why not properly captialize instead of relaying on this filter?

arikfr commented on client/app/visualizations/chart/index.js:
```diff
@@ -77,7 +77,13 @@ function ChartEditor(ColorPalette, clientConfig) {
 
-      scope.xAxisScales = ['datetime', 'linear', 'logarithmic', 'category'];
+      scope.xAxisScales = [
+        { label: 'default', value: '-' },
```
Maybe `Auto Detect` is a better description?

emtwo commented:
@arikfr I addressed your comments. In terms of the regex concern, please see the comment I left. I updated it to be more restrictive but if we're still not ok with this, perhaps we can remove it entirely from this PR. For some context, the motivation for that change originally came from this issue in the mozilla fork: https://github.com/mozilla/redash/issues/338.

arikfr commented:
@emtwo thank you for the updates! As for the regex, while it limits the range, it can still wrongly detect large numbers as dates. As this is a Mozilla specific issue, I would prefer that you address it in some other way.

Some ideas --

1. Change the format in the query itself (if it's frequently used, maybe create a view?), so it's properly loaded as a date value.
2. Add some post processing code in the query runner(s) to update the format of these columns to a more universal one.

In the longer term, it can be great if the user could define the field types of the result set. Similar to how we do it now for the table visualization, but in a global way that will be used in all the visualizations. But this will take some time to implement properly...

emtwo commented:
@arikfr Indeed this is a fairly Mozilla-specific issue so I have taken that change out of this PR and we can re-evaluate how we'll handle it on our side.

This PR is now only for adding `Auto Detect` for the x-axis type.

arikfr commented:
Thanks!
